modules/homa-lite/src/lib.rs

```diff
@@ -283,36 +373,505 @@ pub mod module {
+
+			XcmUnbondFee::<T>::put(xcm_unbond_fee);
+			Self::deposit_event(Event::<T>::XcmUnbondFeeSet(xcm_unbond_fee));
+			Ok(())
+		}
+
+		/// Put in an request to redeem Staking currencies used to mint Liquid currency.
+		/// The redemption will happen after the currencies are unbonded on the relaychain.
+		///
+		/// Parameters:
+		/// - `liquid_amount`: The amount of liquid currency to be redeemed into Staking currency.
+		/// - `additional_fee`: Percentage of the fee to be awarded to the minter.
+		#[pallet::weight(< T as Config >::WeightInfo::request_redeem())]
+		#[transactional]
+		pub fn request_redeem(origin: OriginFor<T>, liquid_amount: Balance, additional_fee: Permill) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			if liquid_amount.is_zero() {
+				// If the maount is zero, cancel previuos redeem request.
+				if let Some((request_amount, _)) = Self::redeem_requests(&who) {
+					// Unreserve the liquid fee and remove the redeem request.
+					let unreserved = T::Currency::unreserve(T::LiquidCurrencyId::get(), &who, request_amount);
+					ensure!(unreserved.is_zero(), Error::<T>::InsufficientReservedBalances);
+					RedeemRequests::<T>::remove(&who);
+
+					Self::deposit_event(Event::<T>::RedeemRequestCancelled(who, request_amount));
+				}
+			} else {
+				// If there are available_staking_balances, redeem immediately with no additional fee.
+				let mut available_staking_balance = Self::available_staking_balance();
+				let actual_liquid_amount = min(
+					liquid_amount,
+					Self::convert_staking_to_liquid(available_staking_balance)?,
+				);
+
+				if !actual_liquid_amount.is_zero() {
+					// Immediately redeem from the available_staking_balances
+					let actual_staking_amount = Self::convert_liquid_to_staking(actual_liquid_amount)?;
+
+					// Redeem from the available_staking_balances costs no extra fee.
+					T::Currency::deposit(T::StakingCurrencyId::get(), &who, actual_staking_amount)?;
+					let slash_amount = T::Currency::slash(T::LiquidCurrencyId::get(), &who, actual_liquid_amount);
+					ensure!(slash_amount.is_zero(), Error::<T>::InsufficientLiquidBalance);
+
+					// Update the available_staking_balance
+					available_staking_balance = available_staking_balance
+						.checked_sub(actual_staking_amount)
+						.expect("min() ensures actual amount cannot be more than the total; qed");
+					AvailableStakingBalance::<T>::put(available_staking_balance);
+
+					Self::deposit_event(Event::<T>::Redeemed(
+						who.clone(),
+						actual_staking_amount,
+						actual_liquid_amount,
+					));
+				}
+
+				// Unredeemed requests are added to a queue.
+				let liquid_remaining = liquid_amount
+					.checked_sub(actual_liquid_amount)
+					.expect("min() ensures actual amount cannot be more than the original; qed");
+				if !liquid_remaining.is_zero() {
+					// Lock the liquid currency.
+					T::Currency::reserve(T::LiquidCurrencyId::get(), &who, liquid_remaining)?;
```

Comment:
this should only reserve the diff between the new request and old request
and should unreserve if old request is more than new request